### PR TITLE
[JIT] Make jit.trace_module reentrant

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29411 [JIT] Make jit.trace_module reentrant**

Fixes https://github.com/pytorch/pytorch/issues/29367

Differential Revision: [D18380559](https://our.internmc.facebook.com/intern/diff/D18380559)

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -1001,36 +1001,38 @@ def weighted_kernel_sum(self, weight):
...
-        if check_trace:
-            if check_inputs is not None:
-                _check_trace(check_inputs, func, check_trace_method,
-                             check_tolerance, _force_outplace, True, _module_class)
-            else:
-                _check_trace([inputs], func, check_trace_method,
-                             check_tolerance, _force_outplace, True, _module_class)
-
-    torch.jit._trace_module_map = None
+    old_module_map = torch.jit._trace_module_map
```
Comment (suo) [inline torch/jit/__init__.py]: can we add a comment here? So next time we don't have to "wtf" our way through the code again.

**Review by suo — APPROVED** [review]
ok boomer

File: torch/jit/__init__.py (hunk 1 above)
Comment (jamesr66a) [reply torch/jit/__init__.py]: I mean it's pretty self explanatory now. The issue was discovering that this was needed in the first place :p

Comment (SsnL) [reply torch/jit/__init__.py]: the usual pattern to do such things is to only replace the attribute when you finish computing it, i.e., start with a local `_trace_module_map` and only do `torch.jit._trace_module_map = _trace_module_map` at the end.

**Pull request closed without merging**